src/MIDebugEngine/Engine.Impl/ExceptionManager.cs

```diff
@@ -434,11 +473,17 @@ private async Task UpdateCatagory(Guid categoryId, ExceptionCategorySettings cat
             {
                 ExceptionBreakpointStates newCategoryState = updates.NewCategoryState.Value;
                 categorySettings.CategoryState = newCategoryState;
+
+                // remove exception breakpoints before categorySettings.CurrentRules is cleared
+                await _commandFactory.RemoveExceptionBreakpoint(categoryId, categorySettings.CurrentRules.Values);
+
+                // question: why does categorySettings.CurrentRules need to be cleared?
                 categorySettings.CurrentRules.Clear();
 
-                IEnumerable<ulong> breakpointIds = await _commandFactory.SetExceptionBreakpoints(categoryId, null, newCategoryState);
-                if (newCategoryState != ExceptionBreakpointStates.None)
+                // only do a generic catch throw if C++ exceptions category is checked
+                if (IsCppExceptionsCategoryChecked(updates))
```

Comment:
>if (IsCppExceptionsCategoryChecked(updates)) [](http://example.com/codeflow?start=16&length=44)

This should give you the same result as `updates.NewCategoryState.GetValueOrDefault(ExceptionBreakpointStates.None).HasFlag(ExceptionBreakpointStates.BreakThrown)`

Reply:
Using `updates.NewCategoryState.HasValue && updates.NewCategoryState.Value != ExceptionBreakpointStates.None` will NOT isolate the C++ exceptions category is checked case.

More specifically, when C++ exceptions category is checked, `updates.NewCategoryState` is `ExceptionBreakpointStates.BreakThrown` AND every key in `RulesToAdd` has value `BreakUserHandled | BreakThrown`.

`updates.NewCategoryState` is ALSO `ExceptionBreakpointStates.BreakThrown` when "<All C++ Exceptions not in this list>" is checked. BUT every key in `RulesToAdd` only has value `BreakUserHandled`.

In short, the only way to distinguish between checked "C++ Exceptions" (category) and checked "<All C++ Exceptions not in this list>" is to check `RulesToAdd`.

Reply:
You can get rid of `BreakUserUnhandled` if you want -- it isn't useful anymore.

> `updates.NewCategoryState` is ALSO ExceptionBreakpointStates.BreakThrown when "<All C++ Exceptions not in this list>" is checked. BUT every key in RulesToAdd only has value BreakUserHandled.

Any time "<All C++ Exceptions not in this list>" is checked -- you want to break on all C++ exceptions. Even the ones in the list. We can't support 'all but not these'.

Reply:
Okay. Seems like I should remove `IsCppExceptionsCategoryChecked()` then.